src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs

```diff
@@ -134,5 +127,278 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                    return false;
+                }
+
+                diagnostics.AddRangeAndFree(applicableDiagnostics);
+
+                // Prior to C# 10, all types in an interpolated string expression needed to be convertible to `object`. After 10, some types
+                // (such as Span<T>) that are not convertible to `object` are permissible as interpolated string components, provided there
+                // is an applicable TryFormatInterpolationHole that accepts them. To preserve langversion, we therefore make sure all components
+                // are convertible to object if the current langversion is lower than the interpolation feature
+                var needToCheckConversionToObject = !Compilation.IsFeatureEnabled(MessageID.IDS_FeatureImprovedInterpolatedStrings) && diagnostics.AccumulatesDiagnostics;
+                var (objectType, conversionDiagnostics) = needToCheckConversionToObject
+                    ? (GetSpecialType(SpecialType.System_Object, diagnostics, unconvertedInterpolatedString.Syntax), BindingDiagnosticBag.GetInstance())
+                    : (null, null);
+
+                // Swap out the first argument of the format calls, which is the (potentially converted) part from the original string with
+                // a placeholder, and put the part into the interpolated string's list of parts.
+                Debug.Assert(builderArguments.Length == unconvertedInterpolatedString.Parts.Length);
+                var interpolatedDataBuilder = ArrayBuilder<MethodArgumentInfo>.GetInstance(builderArguments.Length);
+                var argumentsBuilder = ArrayBuilder<BoundExpression>.GetInstance(3);
+                var partsBuilder = ArrayBuilder<BoundExpression>.GetInstance(builderArguments.Length);
+                int baseStringLength = 0;
+                int numFormatHoles = 0;
+
+                for (int i = 0; i < builderArguments.Length; i++)
+                {
+                    var currentFormatCall = builderArguments[i];
+                    Debug.Assert(currentFormatCall.Arguments.Length > 0);
+                    Debug.Assert(argumentsBuilder.Count == 0);
+
+                    var newPart = currentFormatCall.Arguments[0];
+                    argumentsBuilder.Add(new BoundInterpolatedStringElementPlaceholder(newPart.Syntax, i, newPart.Type));
+                    if (currentFormatCall.Arguments.Length > 1)
+                    {
+                        argumentsBuilder.AddRange(currentFormatCall.Arguments, 1, currentFormatCall.Arguments.Length - 1);
+                    }
+
+                    interpolatedDataBuilder.Add(currentFormatCall with { Arguments = argumentsBuilder.ToImmutableAndClear() });
+
+                    var currentPart = unconvertedInterpolatedString.Parts[i];
+                    if (currentPart is BoundStringInsert insert)
+                    {
+                        numFormatHoles++;
+
+                        if (needToCheckConversionToObject)
```

Comment:
Consider extracting this logic to another local function

Reply:
I'm going to hold off on this for now, as I think this code will need to be updated to get a better error experience for things that cannot be in interpolated string holes (such as `ReadOnlySpan<char>`

---
In reply to: [609046151](https://github.com/dotnet/roslyn/pull/52422#discussion_r609046151) [](ancestors = 609046151)